File: src/generator.test.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { write } from './generator';
import { Session, Message } from './parser/types';

function makeSession(
  tool: string,
  project: string,
  ts: Date,
  ...msgs: Message[]
): Session {
  return { tool, project, timestamp: ts, messages: msgs };
}

describe('generator.write', () => {
  it('should create a file in the correct month subdirectory', () => {
    const outDir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-gen-test-'));
    const ts = new Date('2026-04-06T14:30:00Z');
    const sessions = [
      makeSession(
        'Claude Code',
        '/home/dev/projects/my-app',
        ts,
        { role: 'user', content: 'Hello' },
        { role: 'assistant', content: 'Hi!' }
      ),
    ];

    write(sessions, outDir);

    const monthDir = path.join(outDir, '2026-04');
    const entries = fs.readdirSync(monthDir);
    expect(entries).toHaveLength(1);
    expect(entries[0]).toMatch(/^2026-04-06_/);
    expect(entries[0]).toMatch(/\.md$/);

    fs.rmSync(outDir, { recursive: true });
  });

  it('should write correct Markdown content with frontmatter', () => {
    const outDir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-gen-test-'));
    const ts = new Date('2026-04-06T14:30:00Z');
    const sessions = [
      makeSession(
        'Cursor',
        '/Users/dev/projects/backend-api',
        ts,
        { role: 'user', content: 'Write a python script to reverse a string.' },
        {
          role: 'assistant',
          content: 'Here is the code:\n```python\ndef reverse_string(s):\n    return s[::-1]\n```',
        }
      ),
    ];

    write(sessions, outDir);

    const monthDir = path.join(outDir, '2026-04');
    const entries = fs.readdirSync(monthDir);
    const content = fs.readFileSync(path.join(monthDir, entries[0]), 'utf8');

    expect(content).toContain('tool: "Cursor"');
    expect(content).toContain('project:');
    expect(content).toContain('timestamp:');
    expect(content).toContain('## User');
    expect(content).toContain('## Assistant');
    expect(content).toContain('reverse a string');
    expect(content).toContain('reverse_string');

    fs.rmSync(outDir, { recursive: true });
  });

  it('should be idempotent (write twice without error)', () => {
    const outDir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-gen-test-'));
    const ts = new Date('2026-04-07T09:00:00Z');
    const sessions = [
      makeSession('Claude Code', '/home/dev/project', ts, {
        role: 'user',
        content: 'First run',
      }),
    ];

    expect(() => write(sessions, outDir)).not.toThrow();
    expect(() => write(sessions, outDir)).not.toThrow();

    fs.rmSync(outDir, { recursive: true });
  });

  it('should write multiple sessions to separate files', () => {
    const outDir = fs.mkdtempSync(path.join(os.tmpdir(), 'histd-gen-test-'));
    const ts = new Date('2026-04-08T10:00:00Z');
    const sessions = [
      makeSession('Cursor', '/projects/alpha', ts, { role: 'user', content: 'alpha question' }),
      makeSession('Claude Code', '/projects/beta', ts, { role: 'user', content: 'beta question' }),
    ];

    write(sessions, outDir);

    const monthDir = path.join(outDir, '2026-04');
    const entries = fs.readdirSync(monthDir);
    expect(entries).toHaveLength(2);

    fs.rmSync(outDir, { recursive: true });
  });
});

File: internal/generator/generator_test.go
package generator_test

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/inevolin/histd/internal/generator"
	"github.com/inevolin/histd/internal/parser"
)

func makeSession(tool, project string, ts time.Time, msgs ...parser.Message) parser.Session {
	return parser.Session{
		Tool:      tool,
		Project:   project,
		Timestamp: ts,
		Messages:  msgs,
	}
}

func TestWriteCreatesFile(t *testing.T) {
	outDir := t.TempDir()
	ts := time.Date(2026, 4, 6, 14, 30, 0, 0, time.UTC)
	sessions := []parser.Session{
		makeSession("Claude Code", "/home/dev/projects/my-app", ts,
			parser.Message{Role: "user", Content: "Hello"},
			parser.Message{Role: "assistant", Content: "Hi!"},
		),
	}
	if err := generator.Write(sessions, outDir); err != nil {
		t.Fatalf("Write() error: %v", err)
	}

	// Expect a file in the month subdirectory.
	monthDir := filepath.Join(outDir, "2026-04")
	entries, err := os.ReadDir(monthDir)
	if err != nil {
		t.Fatalf("reading month directory: %v", err)
	}
	if len(entries) != 1 {
		t.Fatalf("expected 1 file, got %d", len(entries))
	}
	name := entries[0].Name()
	if !strings.HasPrefix(name, "2026-04-06_") {
		t.Errorf("filename %q should start with 2026-04-06_", name)
	}
	if !strings.HasSuffix(name, ".md") {
		t.Errorf("filename %q should end with .md", name)
	}
}

func TestWriteMarkdownContent(t *testing.T) {
	outDir := t.TempDir()
	ts := time.Date(2026, 4, 6, 14, 30, 0, 0, time.UTC)
	sessions := []parser.Session{
		makeSession("Cursor", "/Users/dev/projects/backend-api", ts,
			parser.Message{Role: "user", Content: "Write a python script to reverse a string."},
			parser.Message{Role: "assistant", Content: "Here is the code:\n```python\ndef reverse_string(s):\n    return s[::-1]\n```"},
		),
	}
	if err := generator.Write(sessions, outDir); err != nil {
		t.Fatalf("Write() error: %v", err)
	}

	// Read the written file.
	monthDir := filepath.Join(outDir, "2026-04")
	entries, _ := os.ReadDir(monthDir)
	content, err := os.ReadFile(filepath.Join(monthDir, entries[0].Name()))
	if err != nil {
		t.Fatal(err)
	}
	md := string(content)

	// Check frontmatter.
	if !strings.Contains(md, "tool: \"Cursor\"") {
		t.Error("missing tool frontmatter")
	}
	if !strings.Contains(md, "project:") {
		t.Error("missing project frontmatter")
	}
	if !strings.Contains(md, "timestamp:") {
		t.Error("missing timestamp frontmatter")
	}
	// Check conversation headers.
	if !strings.Contains(md, "## User") {
		t.Error("missing ## User header")
	}
	if !strings.Contains(md, "## Assistant") {
		t.Error("missing ## Assistant header")
	}
	// Check content.
	if !strings.Contains(md, "reverse a string") {
		t.Error("user message content missing")
	}
	if !strings.Contains(md, "reverse_string") {
		t.Error("assistant message content missing")
	}
}

func TestWriteIdempotent(t *testing.T) {
	outDir := t.TempDir()
	ts := time.Date(2026, 4, 7, 9, 0, 0, 0, time.UTC)
	sessions := []parser.Session{
		makeSession("Claude Code", "/home/dev/project", ts,
			parser.Message{Role: "user", Content: "First run"},
		),
	}

	// Write twice; should not error.
	if err := generator.Write(sessions, outDir); err != nil {
		t.Fatalf("first Write() error: %v", err)
	}
	if err := generator.Write(sessions, outDir); err != nil {
		t.Fatalf("second Write() error: %v", err)
	}
}

func TestWriteMultipleSessions(t *testing.T) {
	outDir := t.TempDir()
	ts := time.Date(2026, 4, 8, 10, 0, 0, 0, time.UTC)
	sessions := []parser.Session{
		makeSession("Cursor", "/projects/alpha", ts,
			parser.Message{Role: "user", Content: "alpha question"},
		),
		makeSession("Claude Code", "/projects/beta", ts,
			parser.Message{Role: "user", Content: "beta question"},
		),
	}
	if err := generator.Write(sessions, outDir); err != nil {
		t.Fatalf("Write() error: %v", err)
	}
	monthDir := filepath.Join(outDir, "2026-04")
	entries, _ := os.ReadDir(monthDir)
	if len(entries) != 2 {
		t.Errorf("expected 2 files, got %d", len(entries))
	}
}
